infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -325,14 +325,79 @@ private void applyShadowStream(UUID streamId, long snapshot) {
         }
     }
 
+    /**
+     * Get the shadow stream's name of given stream id, note that we are now using stream id instead
+     * of stream name as part of shadow stream's name.
+     *
+     * @param regularStreamId Given stream id
+     * @return Corresponding shadow stream's name
+     */
+    private String getShadowStreamName(UUID regularStreamId) {
+        return regularStreamId + SHADOW_STREAM_SUFFIX;
+    }
+
+    /**
+     * Get the shadow stream's UUID of an original stream from its UUID.
+     *
+     * @param regularStreamId UUID of the original stream
+     * @return UUID of the original stream's shadow stream
+     */
+    private UUID getShadowStreamUUID(UUID regularStreamId) {
+        return CorfuRuntime.getStreamID(getFullyQualifiedTableName(CORFU_SYSTEM_NAMESPACE,
+                getShadowStreamName(regularStreamId)));
+    }
+
+    /**
+     * Rebuild set of streams to replicate, and the stream to tags map from registry table, and
+     * update them to LogReplicationConfig. This method will be invoked after TRANSFER phase of
+     * snapshot sync and before applying shadow streams to their original streams.
+     */
+    private void updateLogReplicationConfig() {
+        CorfuTable<TableName, CorfuRecord<TableDescriptors, TableMetadata>> registryTable =
+                rt.getTableRegistry().getRegistryTable();
+
+        registryTable.forEach((tableName, tableRecord) -> {
+            UUID streamId = CorfuRuntime.getStreamID(getFullyQualifiedTableName(tableName));
+            dataStreamToTagsMap.putIfAbsent(streamId, new HashSet<>());
+            dataStreamToTagsMap.get(streamId).addAll(
+                    tableRecord.getMetadata()
+                            .getTableOptions()
+                            .getStreamTagList()
+                            .stream()
+                            .map(streamTag -> TableRegistry.getStreamIdForStreamTag(tableName.getNamespace(), streamTag))
+                            .collect(Collectors.toList()));
+        });
+        // Get streams to replicate from registry table and update log replication config.
+        Set<UUID> streamsToReplicate = registryTable.entryStream()
+                .filter(entry -> entry.getValue().getMetadata().getTableOptions().getIsFederated())
+                .map(entry -> CorfuRuntime.getStreamID(getFullyQualifiedTableName(entry.getKey())))
+                .collect(Collectors.toSet());
+        // Update the dataStreamToTagsMap in LogReplicationConfig, clear the existing mapping as
+        // this is the point that dataStreamToTagsMap should be initialized in LogReplicationConfig
+        config.updateDataStreamToTagsMap(dataStreamToTagsMap, true);
+        config.updateStreamsToReplicateStandby(streamsToReplicate, true);
```

Comment:
what is the difference between L378 and 379? cant the 2 operations be consolidated?  this can be confusing and error-prone.